src/ast/passes/fold_literals.cpp

```diff
@@ -660,14 +674,30 @@ std::optional<Expression> LiteralFolder::visit(BlockExpr &expr)
 {
   Visitor<LiteralFolder, std::optional<Expression>>::visit(expr);
 
-  // We fold this only if the statment list is empty, and we find a literal
+  // We fold this only if the statement list is empty, and we find a literal
   // as the expression value.
   if (expr.stmts.empty() &&
       (expr.expr.is<Integer>() || expr.expr.is<NegativeInteger>() ||
        expr.expr.is<Boolean>() || expr.expr.is<String>())) {
     return expr.expr;
   }
 
+  StatementList stmt_list;
+  for (auto &stmt : expr.stmts) {
+    visit(stmt);
+    if (auto *while_stmt = stmt.as<While>()) {
+      visit(while_stmt->cond);
+      if (is_literal(while_stmt->cond)) {
+        if (!eval_bool(while_stmt->cond)) {
+          continue;
+        }
```

Comment:
Did you want to inject an error for `while (true)`? Or leave this to a later pass. It feels appropriate because possibly we are trying to "unroll" it here as it is a literal?

Reply:
`while(true)` is fine and we don't want to unroll it at all (realized this the hard way) because you can have a `break`, `exit`, `return` inside it (possibly surrounded by some `if` where the conditional state changes with each iteration) and it's working as expected.